pulp/apis/sas_api.py

```diff
@@ -0,0 +1,623 @@
+from io import StringIO
+from contextlib import redirect_stdout
+import os
+import sys
+from .. import constants
+import warnings
+from typing import Union
+
+from uuid import uuid4
+
+# The maximum length of the names of variables and constraints.
+MAX_NAME_LENGTH=256
+
+# This combines all status codes from OPTLP/solvelp and OPTMILP/solvemilp
+SOLSTATUS_TO_STATUS = {
+    "OPTIMAL": constants.LpStatusOptimal,
+    "OPTIMAL_AGAP": constants.LpStatusOptimal,
+    "OPTIMAL_RGAP": constants.LpStatusOptimal,
+    "OPTIMAL_COND": constants.LpStatusOptimal,
+    "TARGET": constants.LpStatusOptimal,
+    "CONDITIONAL_OPTIMAL": constants.LpStatusOptimal,
+    "FEASIBLE": constants.LpStatusNotSolved,
+    "INFEASIBLE": constants.LpStatusInfeasible,
+    "UNBOUNDED": constants.LpStatusUnbounded,
+    "INFEASIBLE_OR_UNBOUNDED": constants.LpStatusUnbounded,
+    "SOLUTION_LIM": constants.LpStatusNotSolved,
+    "NODE_LIM_SOL": constants.LpStatusNotSolved,
+    "NODE_LIM_NOSOL": constants.LpStatusNotSolved,
+    "ITERATION_LIMIT_REACHED": constants.LpStatusNotSolved,
+    "TIME_LIM_SOL": constants.LpStatusNotSolved,
+    "TIME_LIM_NOSOL": constants.LpStatusNotSolved,
+    "TIME_LIMIT_REACHED": constants.LpStatusNotSolved,
+    "ABORTED": constants.LpStatusNotSolved,
+    "ABORT_SOL": constants.LpStatusNotSolved,
+    "ABORT_NOSOL": constants.LpStatusNotSolved,
+    "OUTMEM_SOL": constants.LpStatusNotSolved,
+    "OUTMEM_NOSOL": constants.LpStatusNotSolved,
+    "FAILED": constants.LpStatusNotSolved,
+    "FAIL_SOL": constants.LpStatusNotSolved,
+    "FAIL_NOSOL": constants.LpStatusNotSolved,
+    "ERROR": constants.LpStatusNotSolved,
+}
+
+CAS_OPTION_NAMES = [
+    "hostname",
+    "port",
+    "username",
+    "password",
+    "session",
+    "locale",
+    "name",
+    "nworkers",
+    "authinfo",
+    "protocol",
+    "path",
+    "ssl_ca_list",
+    "authcode",
+]
+
+
+class SAS94(LpSolver):
+    name = "SAS94"
+
+    try:
+        global saspy
+        import saspy
+
+    except ImportError:
+
+        def available(self):
+            """True if SAS94 is available."""
+            return False
+
+        def actualSolve(self, lp, callback=None):
+            """Solves a well-formulated lp problem."""
+            raise PulpSolverError("SAS94 : Not Available")
+
+    else:
+        saspy.logger.setLevel(log.level)
+        def __init__(
+            self,
+            mip=True,
+            msg=True,
+            warmStart=False,
+            **solverParams,
+        ):
+            """
+            :param bool mip: if False, assume LP even if integer variables
+            :param bool msg: if False, no log is shown
+            :param bool warmStart: if False, no warmstart or initial primal solution provided.
+            """
+            LpSolver.__init__(
+                self,
+                mip=mip,
+                msg=msg,
+            )
+            self._sas = saspy.SASsession()
+            self.warmStart=warmStart
```

Comment:
`warmstart` can be passed to `LpSolver` and then becomes available inside `self.optionsDict`. See how it's used in other solvers. No need to assign it to the object.